File: src/llm.py
import asyncio
import base64
import random
from dataclasses import dataclass

from openai import AsyncOpenAI

from .config import OPENROUTER_API_KEY, MODELS, IMAGE_MODEL, VIDEO_MODEL, VOICE_MODEL
from .utils import timed

SUMMARIZATION_PROMPT = """Ты — бот-саммаризатор сообщений в Telegram.

Сообщения поступают в формате:
```
### ID Name
  text
```

Перескажи самые интересные / смешные моменты. 

Требования:
0. Язык ответа — русский
1. Длина — приблизительно до 1200 символов
2. Пиши только сам пересказ! Без фразы "Вот основные моменты", без заголовка "Пересказ", без рассуждений о чате в целом.
3. Для форматирования используй html (не markdown).
4. Используй только теги, поддерживаемые Telegram:
   - <b>текст</b> (жирный)
   - <i>текст</i> (курсив)
   - <a href="URL">текст</a> (ссылки)
5. Вместо списков (<ul>) используй символы-буллеты (• или -) и обычный перенос строки (\\n).
6. Обязательно закрывай все теги."""

openai_client = AsyncOpenAI(
    api_key=OPENROUTER_API_KEY,
    base_url="https://openrouter.ai/api/v1",
)


def _indent(text: str) -> str:
    return "\n".join(f"  {line}" for line in text.split("\n"))


def _format_attachment_block(attachment: dict | None, description: str | None = None) -> str | None:
    if not attachment:
        return None
    
    att_type = attachment.get("type")
    
    if att_type == "photo":
        if description:
            return f"<photo>\n{_indent(description)}\n</photo>"
        return "<photo />"
    
    if att_type == "voice":
        if description:
            return f"<voice>\n{_indent(description)}\n</voice>"
        return "<voice />"
    
    if att_type == "video_note":
        if description:
            return f"<video_note>\n{_indent(description)}\n</video_note>"
        return "<video_note />"
    
    if att_type == "video":
        return "<video />"
    
    if att_type == "animation":
        return "<gif />"
    
    if att_type == "sticker":
        if description:
            return f"<sticker>\n{_indent(description)}\n</sticker>"
        emoji = attachment.get("emoji", "")
        if emoji:
            return f"<sticker>{emoji}</sticker>"
        return "<sticker />"
    
    if att_type == "document":
        file_name = attachment.get("file_name", "файл")
        return f"<document>{file_name}</document>"
    
    if att_type == "poll":
        question = attachment.get("question", "")
        options = attachment.get("options", [])
        if options:
            options_str = "\n".join(f"  - {opt}" for opt in options)
            return f"<poll>{question}\n{options_str}\n</poll>"
        return f"<poll>{question}</poll>"
    
    if att_type == "location":
        return "<location />"
    
    if att_type == "new_members":
        names = attachment.get("names", "")
        return f"<new_members>{names}</new_members>"
    
    return None


def _get_attachment_description(msg_data: dict) -> str | None:
    attachment = msg_data.get("attachment")
    if attachment:
        return attachment.get("description")
    return None


def format_message_for_prompt(msg_data: dict) -> str:
    msg_id = msg_data["message_id"]
    name = msg_data["sender_name"]
    text = msg_data.get("text", "")
    reply_to = msg_data.get("reply_to_message_id")
    attachment = msg_data.get("attachment")
    forward_name = msg_data.get("forward_sender_name")
    
    labels = []
    if reply_to:
        labels.append(f"reply to {reply_to}")
    if forward_name:
        labels.append(f"forward from {forward_name}")
    
    labels_str = f" [{', '.join(labels)}]" if labels else ""
    
    parts = [f"### {msg_id} {name}{labels_str}"]
    
    if text:
        parts.append(_indent(text))
    
    attachment_block = _format_attachment_block(attachment, _get_attachment_description(msg_data))
    if attachment_block:
        parts.append(attachment_block)
    
    return "\n".join(parts)


def generate_full_prompt(messages: list[dict]) -> str:
    messages_text = "\n\n".join(format_message_for_prompt(m) for m in messages)
    return SUMMARIZATION_PROMPT + f"\n\n<messages>\n{messages_text}\n</messages>"


@dataclass
class SummaryResult:
    text: str
    model: str
    input_tokens: int | None = None
    output_tokens: int | None = None
    cost: float | None = None


async def generate_summary(messages: list[dict]) -> SummaryResult:
    model = random.choice(MODELS)
    prompt = generate_full_prompt(messages)
    
    response = await openai_client.responses.create(
        model=model,
        input=prompt
    )
    
    result = SummaryResult(text=response.output_text, model=model)
    result.cost = extract_cost(response)
    
    if usage := getattr(response, 'usage', None):
        result.input_tokens = getattr(usage, 'input_tokens', None)
        result.output_tokens = getattr(usage, 'output_tokens', None)
    
    return result


@dataclass
class DescribeResult:
    text: str
    cost: float | None = None


def get_model_short_name(model: str) -> str:
    # 'anthropic/claude-opus-4.5' -> 'claude-opus-4.5'
    return model.split('/')[-1]


def extract_cost(response) -> float | None:
    if usage := getattr(response, 'usage', None):
        return getattr(usage, 'cost', None)
    return None


async def call_multimodal(model: str, prompt: str, media_content: dict) -> DescribeResult:
    response = await openai_client.responses.create(
        model=model,
        input=[{
            'role': 'user',
            'content': [
                {'type': 'input_text', 'text': prompt},
                media_content,
            ]
        }]
    )
    return DescribeResult(text=response.output_text, cost=extract_cost(response))


async def describe_image(base64_image: str) -> DescribeResult:
    return await call_multimodal(IMAGE_MODEL, 'Что изображено на картинке? Кратко',
        {'type': 'input_image', 'image_url': f'data:image/jpeg;base64,{base64_image}'})


async def describe_sticker(base64_image: str) -> DescribeResult:
    return await call_multimodal(IMAGE_MODEL,
        'Очень кратко опиши стикер. Если стикер представляет собой скриншот сообщения, ответь в формате "Имя:\\nтекст сообщения"',
        {'type': 'input_image', 'image_url': f'data:image/jpeg;base64,{base64_image}'})


async def describe_video_note(base64_video: str) -> DescribeResult:
    return await call_multimodal(VIDEO_MODEL, 'Что происходит / какие слова говорятся в видеосообщении?',
        {'type': 'input_video', 'video_url': f'data:video/mp4;base64,{base64_video}'})


@timed
async def convert_ogg_to_mp3(audio_bytes: bytes) -> bytes:
    proc = await asyncio.create_subprocess_exec(
        'ffmpeg', '-loglevel', 'error', '-i', 'pipe:0', '-f', 'mp3', 'pipe:1',
        stdin=asyncio.subprocess.PIPE,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
    )
    stdout, stderr = await proc.communicate(audio_bytes)
    if proc.returncode != 0:
        raise RuntimeError(f"ffmpeg error: {stderr.decode()}")
    return stdout


async def describe_voice(audio_bytes: bytes) -> DescribeResult:
    """Конвертирует ogg в mp3 — Responses API не принимает ogg."""
    mp3_bytes = await convert_ogg_to_mp3(audio_bytes)
    base64_audio = base64.b64encode(mp3_bytes).decode()

    return await call_multimodal(
        VOICE_MODEL,
        'Расшифруй это голосовое сообщение. Выведи только текст.',
        {'type': 'input_audio', 'input_audio': {'data': base64_audio, 'format': 'mp3'}},
    )

File: src-ts/llm.ts
import OpenAI from 'openai';
import { OPENROUTER_API_KEY, MODELS, IMAGE_MODEL, VIDEO_MODEL, VOICE_MODEL } from './config.ts';
import { logWarning, timed } from './utils.ts';

const CONTENT_FILTER_MESSAGE = '[тема слишком опасна]';

const SUMMARIZATION_PROMPT = `Ты — бот-саммаризатор сообщений в Telegram.

Сообщения поступают в формате:
\`\`\`
### ID Name
  text
\`\`\`

Перескажи самые интересные / смешные моменты.

Требования:
0. Язык ответа — русский
1. Длина — приблизительно до 1200 символов
2. Пиши только сам пересказ! Без фразы "Вот основные моменты", без заголовка "Пересказ", без рассуждений о чате в целом.
3. Для форматирования используй html (не markdown).
4. Используй только теги, поддерживаемые Telegram:
   - <b>текст</b> (жирный)
   - <i>текст</i> (курсив)`;

const openai = new OpenAI({
  apiKey: OPENROUTER_API_KEY,
  baseURL: 'https://openrouter.ai/api/v1',
});

function indent(text: string): string {
  return text.split('\n').map(l => `  ${l}`).join('\n');
}

function formatAttachmentBlock(attachment: Record<string, unknown> | null | undefined, description?: string | null): string | null {
  if (!attachment) return null;
  const type = attachment.type as string;

  if (type === 'photo') return description ? `<photo>\n${indent(description)}\n</photo>` : '<photo />';
  if (type === 'voice') return description ? `<voice>\n${indent(description)}\n</voice>` : '<voice />';
  if (type === 'video_note') return description ? `<video_note>\n${indent(description)}\n</video_note>` : '<video_note />';
  if (type === 'video') return '<video />';
  if (type === 'animation') return '<gif />';
  if (type === 'sticker') {
    if (description) return `<sticker>\n${indent(description)}\n</sticker>`;
    const emoji = attachment.emoji as string | undefined;
    return emoji ? `<sticker>${emoji}</sticker>` : '<sticker />';
  }
  if (type === 'document') return `<document>${attachment.file_name ?? 'файл'}</document>`;
  if (type === 'poll') {
    const question = attachment.question as string ?? '';
    const options = attachment.options as string[] ?? [];
    if (options.length) return `<poll>${question}\n${options.map(o => `  - ${o}`).join('\n')}\n</poll>`;
    return `<poll>${question}</poll>`;
  }
  if (type === 'location') return '<location />';
  if (type === 'new_members') return `<new_members>${attachment.names ?? ''}</new_members>`;
  return null;
}

export interface MessageData {
  message_id: number;
  sender_name: string;
  text?: string;
  reply_to_message_id?: number | null;
  forward_sender_name?: string | null;
  attachment?: Record<string, unknown> | null;
}

export function formatMessageForPrompt(msg: MessageData): string {
  const labels: string[] = [];
  if (msg.reply_to_message_id) labels.push(`reply to ${msg.reply_to_message_id}`);
  if (msg.forward_sender_name) labels.push(`forward from ${msg.forward_sender_name}`);

  const labelsStr = labels.length ? ` [${labels.join(', ')}]` : '';
  const parts = [`### ${msg.message_id} ${msg.sender_name}${labelsStr}`];

  if (msg.text) parts.push(indent(msg.text));

  const description = (msg.attachment as Record<string, unknown> | null | undefined)?.description as string | null | undefined;
  const block = formatAttachmentBlock(msg.attachment, description);
  if (block) parts.push(block);

  return parts.join('\n');
}

export function generateFullPrompt(messages: MessageData[]): string {
  const messagesText = messages.map(formatMessageForPrompt).join('\n\n');
  return `${SUMMARIZATION_PROMPT}\n\n<messages>\n${messagesText}\n</messages>`;
}

export interface SummaryResult {
  text: string;
  model: string;
  input_tokens: number | null;
  output_tokens: number | null;
  cost: number | null;
}

function extractCost(response: { usage?: unknown }): number | null {
  return (response.usage as Record<string, unknown> | null | undefined)?.cost as number ?? null;
}

export async function generateSummary(messages: MessageData[]): Promise<SummaryResult> {
  const model = MODELS[Math.floor(Math.random() * MODELS.length)];
  const prompt = generateFullPrompt(messages);

  const response = await openai.responses.create({ model, input: prompt });

  if (response.incomplete_details?.reason === 'content_filter') {
    throw new Error(`content_filter: ${model}`);
  }

  return {
    text: response.output_text,
    model,
    input_tokens: response.usage?.input_tokens ?? null,
    output_tokens: response.usage?.output_tokens ?? null,
    cost: extractCost(response),
  };
}

export interface DescribeResult {
  text: string;
  cost: number | null;
}

async function callMultimodal(model: string, prompt: string, mediaContent: unknown): Promise<DescribeResult> {
  const response = await openai.responses.create({
    model,
    input: [{
      role: 'user',
      content: [
        { type: 'input_text', text: prompt },
        mediaContent,
      ],
    }] as Parameters<typeof openai.responses.create>[0]['input'],
  });
  if (response.incomplete_details?.reason === 'content_filter') {
    logWarning(`content_filter: ${model}`);
    return { text: CONTENT_FILTER_MESSAGE, cost: extractCost(response) };
  }
  return { text: response.output_text, cost: extractCost(response) };
}

export function describeImage(base64Image: string): Promise<DescribeResult> {
  return callMultimodal(IMAGE_MODEL, 'Что изображено на картинке? Кратко',
    { type: 'input_image', image_url: `data:image/jpeg;base64,${base64Image}` });
}

export function describeSticker(base64Image: string): Promise<DescribeResult> {
  return callMultimodal(IMAGE_MODEL,
    'Очень кратко опиши стикер. Если стикер представляет собой скриншот сообщения, ответь в формате "Имя:\\nтекст сообщения"',
    { type: 'input_image', image_url: `data:image/jpeg;base64,${base64Image}` });
}

export function describeVideoNote(base64Video: string): Promise<DescribeResult> {
  return callMultimodal(VIDEO_MODEL, 'Что происходит / какие слова говорятся в видеосообщении (если говорятся)?',
    { type: 'input_video', video_url: `data:video/mp4;base64,${base64Video}` });
}

export const convertOggToMp3 = timed('convert_ogg_to_mp3', async (audioBytes: Uint8Array): Promise<Buffer> => {
  return new Promise((resolve, reject) => {
    const { spawn } = require('child_process');
    const ffmpeg = spawn('ffmpeg', ['-loglevel', 'error', '-i', 'pipe:0', '-f', 'mp3', 'pipe:1']);
    const chunks: Buffer[] = [];
    ffmpeg.stdout.on('data', (chunk: Buffer) => chunks.push(chunk));
    ffmpeg.stderr.on('data', (data: Buffer) => {
      // Collected if needed
    });
    ffmpeg.on('close', (code: number) => {
      if (code !== 0) reject(new Error(`ffmpeg exited with code ${code}`));
      else resolve(Buffer.concat(chunks));
    });
    ffmpeg.on('error', reject);
    ffmpeg.stdin.write(audioBytes);
    ffmpeg.stdin.end();
  });
});

export async function describeVoice(audioBytes: Uint8Array): Promise<DescribeResult> {
  const mp3 = await convertOggToMp3(audioBytes);
  const base64Audio = mp3.toString('base64');
  return callMultimodal(VOICE_MODEL,
    'Расшифруй это голосовое сообщение. Выведи только текст.',
    { type: 'input_audio', input_audio: { data: base64Audio, format: 'mp3' } });
}

const OBVIOUS_PROVIDERS = new Set(['google', 'anthropic', 'deepseek', 'qwen', 'x-ai']);

export function getModelShortName(model: string): string {
  const [provider, modelShort] = model.split('/');
  if (modelShort && OBVIOUS_PROVIDERS.has(provider)) return modelShort;
  return model;
}
